include/oneapi/dnnl/dnnl.hpp

```diff
@@ -3392,6 +3410,24 @@ struct memory : public handle<dnnl_memory_t> {
         reset(result);
     }
 
+    /// Constructs a memory object that wraps a host scalar value.
+    /// @note The scalar value is copied into the newly allocated memory storage,
+    ///     so the user does not need to manage the lifetime of the original scalar data.
+    ///
+    /// @tparam T Type of the scalar value.
+    /// @param md The memory descriptor that defines data type.
+    /// @param value The scalar value to be wrapped by the memory object.
+    ///
+    /// @throws error if the memory object could not be created.
+    template <typename T>
+    memory(const desc &md, const T &value) {
```

Comment:
One potential pitfall is that `T` may be different from the data type contained in md. It should be an user fault but it propagates to the library through the C API when casting the input to `void *`.

Reply:
> It should be an user fault

That's exactly the reason why I think it's not a problem :) User may also pass a handle to a buffer whose size is different from the one described in md. It's all the user's responsibility.

We could add a check to play it safe though:
```cpp
switch (md.get_data_type()) {
    case memory::data_type::f32: error::wrap_c_api(
        std::is_same<T, float>::value ? dnnl_success : dnnl_invalid_arguments, "Data type mismatch");
    break;
    default: error::wrap_c_api(dnnl_invalid_arguments, "Unknown data type");
}
```

Reply:
Host-side scalars are required for both scales and zp, I think checking each possible type with switch()/case is probably too much (but I could do so if necessary)? And just a check for floating point is not enough due to possible integer values. Wouldn't a check for data type size be sufficient/most meaningful (and simple enough)? I've added one in [0b498d7](https://github.com/uxlfoundation/oneDNN/pull/3506/commits/0b498d7f97fff889853b83e49ecd6bc2498f0668).